src/analysis/souffle/taint.dl

```diff
@@ -71,6 +95,37 @@ mayHaveTag(tgt, owner, tag) :- hasTag(tgt, owner, tag).
 mayHaveTag(tgt, owner, tag) :-
     resolvedEdge(owner, src, tgt), mayHaveTag(src, owner, tag), !removeTag(tgt, owner, tag).
 
+// Integrity tag rules.
+
+isIntegTag(integTag) :- directHasIntegTag(_, integTag).
+
+// The inverse integrity tag is the same text as the integrity tag except with
+// a '!' in front of it.
+integTagToInv(integTag, cat("!", integTag)) :- isIntegTag(integTag).
+
+// Introduce an inverse integrity tag at any access point that has no incoming
+// edges and is not directly given the corresponding integrity tag.
+mayHaveInvIntegTag(ap, invIntegTag) :-
+  isAccessPath(ap), integTagToInv(integTag, invIntegTag), !edge(_, ap),
+  !directHasIntegTag(ap, integTag).
+
```

Comment:
Might be simpler to do it as follows?

```suggestion
.decl lacksIntegrity(ap: AccessPath, owner: Principal, tag: Tag)

lacksIntegrity(ap, owner, tag) :-
  isAccessPath(ap), isPrincipal(owner), edge(_, ap),
  lacksIntegrity(ap, owner, tag).
```

Reply:
We may also need to scope out the `tag` in the second rule:

```
lacksIntegrity(ap, owner, tag) :- 
	 isAccessPath(ap), isPrincipal(owner), isIntegrityTag(tag), !directHasInteg(ap, owner, tag).
```

Reply:
I don't think the above will work. The second rule:

```Datalog
lacksIntegrity(ap, owner, tag) :- isAccessPath(ap), isPrincipal(owner), !directHasInteg(ap, owner, tag).
```

will make every `AccessPath` that is not directly given an `IntegrityTag` lack that `IntegrityTag`. We want `AccessPath`s not directly given an `IntegrityTag` but for which all inputs have an `IntegrityTag` to have the `IntegrityTag`. This is why my current rules only introduce inverse integrity tags on `AccessPath`s not directly given an `IntegrityTag` that have no incoming edges: there is no way for them to get an `IntegrityTag`, and thus the integrity tag must be lacking. From there, I can propagate the lack of integrity tags like confidentiality tags.

Reply:
/\ I think as I was typing that comment just the first message was in the thread, so what I wrote might be out of sync.

Reply:
From another comment thread from @markww :
> We want AccessPaths not directly given an IntegrityTag but for which all inputs have an IntegrityTag to have the IntegrityTag

Ah, now that you call this out, I am not sure if this is always desirable. Consider something like `TimestampMillis` and a particle with two inputs (say 'a' and 'b') and one output (say 'a + b'). Suppose that the inputs have `TimestampMillis` doesn't mean that the output should have the tag. That said, if we we think of integrity as provenance (say "screentext"), it would make sense to propagate integrity to all the derived outputs. I think we will want something like `edgePreservesIntegrity("screentext")` and use that in propagation.

In any case, as to your comment about `lacksIntegrity` not working, you are both right. We will have to combine them into one. Here is what I had written up before seeing @aferr's comment: 
```
lacksIntegrity(ap, owner, tag) :-
  isAccessPath(ap), isPrincipal(owner), isIntegrityTag(tag), edge(src, ap),
  lacksIntegrity(src, owner, tag), !directHasInteg(ap, owner, tag).

```

Reply:
Another way we could approach this is by making `IntegrityTag`s *never* propagate automatically. That is, for each `IntegrityTag`, you must specify explicitly the rules that cause, for some set of input nodes, an integrity tag to appear on an output node.

Removing automatic propagation would remove the ambient double-negation that occurs due to propagation and would allow us to only consider `IntegrityTag`s in the affirmative. If we do that, all of the stratification requirements evaporate and we can easily handle the SQL cases we are considering elsewhere.

Reply:
I am supportive of this idea, but we will still need to deal with intersection. Suppose we have a rule like `edgePreservesIntegrity(src, tgt, tag)`. We will still need to make sure that _all_ `edgePreservesIntegrity(src, tgt, tag)` to be true before we set that the integrity of `tgt` to be true.

Reply:
This may not be the right approach, but my idea was that we wouldn't even have a fact like `edgePreservesIntegrity`. Instead, we would always use policy rules that may match on certain subsections of the Raksha graph to propagate integrity tags.

I think we could, at least in most cases, sidestep intersection due to specific knowledge about the shape of the graph encoded within these propagation rules. Consider, for instance, the example where we remove the lower bits of a timestamp to prevent fingerprinting a user. We could write such a rule like this:

```Datalog
hasIntegrityTag(tgt, "TimestampSeconds") :-
  isOperation(tgt, "/"), // Indicates that this access path holds the result of a divide
  isDividendOf(src1, tgt), // Indicates that this access path holds the value being divided at tgt
  isShiftWidthOf(src2, tgt), // Indicates that this access path holds the value being divided by at tgt
  hasIntegrityTag(src1, "TimestampMillis"), // Is this a millisecond value.
  hasIntegrityTag(src2, "DivisibleBy1000"), // Do we know this definitely has three 0s in the low bits
  count : { edge (_, tgt) } = 2. // Only two edges enter this node.
```

The above rule effectively performs an intersection for the specific case of a division. It manages to do this without negation upon dynamic facts because it knows that a division should only take two inputs. We check that this is the case with the `count` operation; this is currently fine because we don't dynamically add edges, so the contents of the `edge` relation is static (we could even assert this property during IR translation to keep it out of Datalog entirely).

A generic propagation rule must use negation because it has to be generic and thus not know the number of inputs that it must intersect. This requires it to state the intersection as no input having the negation of the property. However, if we know that there are `X` inputs, we can exhaustively say `x1, x2, x3, ...` `X` times.

Of course, this does not work for circumstances where the user does not have specific knowledge of the number of inputs going into an `AccessPath` that is relevant to an integrity property. But perhaps those circumstances are rare. All Arcs `Particle`s know the number of input handles that they have. Most SQL expressions and statements also know the exact number of incoming inputs (the list in `IN` being a notable exception). We need a constraint to make things stratify, and perhaps this is a least-onerous one to use.

Reply:
@bgogul and I had a video call in which he convinced me that automatic propagation is needed in at least some circumstances. We sort of stumbled upon an idea together of having a special binary `phi` operation which actually performs the merging of confidentiality state and integrity state for two incoming `AccessPath`s. A `phi` result has a confidentiality tag when at least one of its inputs has that confidentiality tag, and has an integrity tag when both of its inputs have that integrity tag. `phi`s can be chained together to capture multiple inputs.

What does this do for us? Well, it means that we can perform intersection without negation: checking that an integrity tag is present on all inputs of a `phi` is equivalent to ensuring that it is present on both of its operands.

When combined with a constraint that each `AccessPath` is the target of at most one incoming edge, this would force all state merging to occur in operations (`phi` may soon be followed by less exotic operations such as `+`, `/`, etc). That would be a radically different approach than what we currently have in this PR, and would necessitate changes elsewhere in the system, so it's probably not in scope for this particular PR.

Reply:
PR #243 contains a prototype of the `phi`-based solution.

Reply:
With the realization of the `phi`-based solution in a different PR, @bgogul convincing me that automatic propagation is necessary, and me switching us to a `lacksIntegrityTag`-based propagation approach, I think everything in this thread has been addressed.